client/core/core.go

```diff
@@ -5146,106 +5214,164 @@ func (c *Core) prepareTrackedTrade(dc *dexConnection, form *TradeForm, crypter e
 			ToVersion:          wallets.toAsset.Version,
 			Options:            form.Options,
 			RedemptionReserves: redemptionReserves,
 			ChangeCoin:         changeID,
-		},
-		Order: ord,
-	}
-	err = c.db.UpdateOrder(dbOrder)
-	if err != nil {
-		logAbandon(fmt.Sprintf("failed to store order in database: %v", err))
-		return nil, nil, fmt.Errorf("Order abandoned due to database error: %w", err)
-	}
+		}
 
-	// Prepare and store the tracker and get the core.Order to return.
-	tracker := newTrackedTrade(dbOrder, preImg, dc, dc.marketEpochDuration(mktID), c.lockTimeTaker, c.lockTimeMaker,
-		c.db, c.latencyQ, wallets, coins, c.notify, c.formatDetails, form.Options, redemptionReserves, refundReserves)
+		// failedOrderError avail the provided error for order error
+		// notification and returns the error.
+		failedOrderError := func(err error) {
+			asyncErrCloser.Done(c.log) // Return all coins before notifying callers via resultChan.
+			if queue {
+				// We only want to send order error notifications if the order
+				// was queued.
+				topic := TopicQueueOrderFailure
+				subject, details := c.formatDetails(topic, meta.TempID.Token(), err)
+				c.notify(newOrderNote(topic, subject, details, db.ErrorLevel, &Order{TempID: meta.TempID.Bytes()}))
+			}
+			resultChan <- &orderSubmissionResult{err: err}
+		}
 
-	tracker.redemptionLocked = tracker.redemptionReserves
-	tracker.refundLocked = tracker.refundReserves
+		if queue {
+			_, err = rand.Read(meta.TempID[:])
+			if err != nil {
+				failedOrderError(fmt.Errorf("unable to generate temporary order ID: %w", err))
+				return
+			}
 
-	if recoveryCoin != nil {
-		tracker.change = recoveryCoin
-		tracker.coinsLocked = false
-		tracker.changeLocked = true
-	}
+			corder := coreOrderFromTrade(ord, meta)
 
-	dc.tradeMtx.Lock()
-	dc.trades[tracker.ID()] = tracker
-	dc.tradeMtx.Unlock()
+			// Store the queued order until a server response is received.
+			dc.queuedOrdersMtx.Lock()
+			dc.queuedOrders[meta.TempID] = corder
+			dc.queuedOrdersMtx.Unlock()
 
-	// Now that the trades map is updated, sync with the preimage request.
-	c.syncOrderPlaced(ord.ID())
+			defer func() {
+				dc.queuedOrdersMtx.Lock()
+				delete(dc.queuedOrders, meta.TempID)
+				dc.queuedOrdersMtx.Unlock()
+			}()
 
-	// Send a low-priority notification.
-	corder := tracker.coreOrder()
-	if !form.IsLimit && !form.Sell {
-		ui := wallets.quoteWallet.Info().UnitInfo
-		subject, details := c.formatDetails(TopicYoloPlaced,
-			ui.ConventionalString(corder.Qty), ui.Conventional.Unit, tracker.token())
-		c.notify(newOrderNote(TopicYoloPlaced, subject, details, db.Poke, corder))
-	} else {
-		rateString := "market"
-		if form.IsLimit {
-			rateString = wallets.trimmedConventionalRateString(corder.Rate)
+			// Return a temp order and send silent note for the queued order.
+			// This improves the UI/UX, so users don't have to wait for orders
+			// especially split tx orders.
+			resultChan <- &orderSubmissionResult{ord: corder}
```

Comment:
How is the caller supposed to receive this unless they get the actual channel from the `return`?

Reply:
With the current design which will eventually change, all callers of `prepareTrackedTrade` get the channel.

Reply:
But that's after the server request has completed, making this fully-synchronous.

Reply:
The server request is launched in a go routine, then callers wait for the result sent via the channel. However, I updated the PR at https://github.com/decred/dcrdex/commit/a439114460eb1e4ba4ef599e33aebc573d92dabe

Reply:
I see. The commit just pushed makes more sense.